datafusion/expr-common/src/type_coercion/binary.rs

```diff
@@ -357,6 +357,14 @@ fn math_decimal_coercion(
         | (Decimal256(_, _), Decimal256(_, _)) => {
             Some((lhs_type.clone(), rhs_type.clone()))
         }
+        // Cross-variant decimal coercion - choose larger variant with appropriate precision/scale
+        (Decimal32(_, _), Decimal64(_, _) | Decimal128(_, _) | Decimal256(_, _))
+        | (Decimal64(_, _), Decimal32(_, _) | Decimal128(_, _) | Decimal256(_, _))
+        | (Decimal128(_, _), Decimal32(_, _) | Decimal64(_, _) | Decimal256(_, _))
+        | (Decimal256(_, _), Decimal32(_, _) | Decimal64(_, _) | Decimal128(_, _)) => {
+            let coerced_type = get_wider_decimal_type_cross_variant(lhs_type, rhs_type)?;
+            Some((coerced_type.clone(), coerced_type))
+        }
```

Comment:
```suggestion
        // Cross-variant decimal coercion - choose larger variant with appropriate precision/scale
        (lhs, rhs) if is_decimal(lhs) && is_decimal(rhs) && std::mem::discriminant(lhs) != std::mem::discriminant(rhs) => {
            let coerced_type = get_wider_decimal_type_cross_variant(lhs_type, rhs_type)?;
            Some((coerced_type.clone(), coerced_type))
        }
```

Using [`is_decimal`](https://github.com/apache/datafusion/blob/2d947b31d0355710dc179d1d72ca5366b7721b2a/datafusion/expr/src/type_coercion/mod.rs#L92-L101) and [`std::mem::discriminant`](https://doc.rust-lang.org/std/mem/fn.discriminant.html)

Dunno if it's better to list out each combination (more explicit but could be easier to miss if we're missing a combination) or do this way which might not be as clear but could be more robust 🤔